### ISSUE TITLE
Deleted and recreated service is ignored by load-balancer

### ISSUE THREAD

**Issue by gordontyler** [rancher/rancher#2561]
I have an application stack containing the following services:

- 30 app services, scaled to 15 instances
- 30 db services, linked to the app services (one-to-one)
- 30 load balancer services, linked to the app services (one-to-one)
- 1 load balancer service, global, linked to the above load balancers

I also have multiple JMeter instances all applying load to different instances of the global load balancer.

I had to remove and re-add a host to the environment. Removing the host caused the containers it was running to be recreated on other hosts -- as expected. After I re-added the host (having recreated the Docker and Rancher install on it), I removed and re-up'd a few of the app services to "rebalance" their containers onto the re-added host.

Subsequently, I have noticed that the request load applied to the global load balancer is not being sent to any of the app services that were recreated. I've tried restarting the individual load balancers for the affected app services. I've tried restarting the global load balancer. I just tried recreating the global load balancer, but there's something going wrong and the rancher-compose up command is not finishing and the service is stuck in "Activating" with all containers running.

**Comment by gordontyler** [comment]
On my second attempt, removing and recreating the global load balancer has completed. Still waiting to see if it's ditributing the load to the recreated app services.

**Comment by gordontyler** [comment]
Yes, I'm seeing load distributed to the recreated app services.

**Comment by deniseschannon** [comment]
Could you share a `docker-compose.yml` and `rancher-compose.yml` so that I could try to reproduce (on a smaller scale)? 

I'm wondering if there was a delay in just getting your containers on the newly added host with the rest of the managed network, so that's what caused the initial request to seem to not hit the new services.

**Comment by gordontyler** [comment]
Sent a Dropbox link to you on the Rancher forums.

**Comment by deniseschannon** [comment]
How many stacks do you have? Would you be able to just click on the **view config** of the stack and send me that docker-compose.yml and rancher-compose.yml?

In the detailed view of the stack, you can click on the paper icon. 
![vvvvvvv](https://cloud.githubusercontent.com/assets/11410997/11072174/7cbcda0c-8799-11e5-84af-1dedf118fa1d.png)

**Comment by gordontyler** [comment]
Clicking that button never finishes loading. :)

I have 40 app services (current scaled to 25 instances each), 40 db services (scale 1), 40 internal load balancer services (scale 1) and a public load balancer (global) pointing at each of the 40 interna; load balancers.

I'll send the compose files that I use to `up` the stack.

**Comment by gordontyler** [comment]
I've put the combined compose files into the same Dropbox folder.

**Comment by deniseschannon** [comment]
@gordontyler Thanks, I"m going to try and simplify a little so that I can test your use case. It might just take me a day or two. :)

**Comment by gordontyler** [comment]
I doubt I'm going to increase the number of services, but I am in the process of increasing the scale of each service. My goal is 50 per service for a total of 2000 instances.

**Comment by deniseschannon** [comment]
Are you still in this bad state? I wonder if the pinging between network agents from the new host to the other hosts) wasn't working before you did all the restarts to get it working again.

**Comment by gordontyler** [comment]
No, I've been able to recover the system.

I see a log of "failed to get ping from agent" messages in the rancher log:

2015-11-10 19:51:52,884 ERROR [:] [] [] [] [ecutorService-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [1] count [4]
2015-11-10 19:51:52,884 ERROR [:] [] [] [] [ecutorService-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [10] count [4]
2015-11-10 19:51:58,010 ERROR [:] [] [] [] [ecutorService-3] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [1] count [5]
2015-11-10 19:51:58,010 ERROR [:] [] [] [] [ecutorService-3] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [10] count [5]

But I'm actively scaling up the services so the load on the hsots may be too high for timely ping responses.

**Comment by deniseschannon** [comment]
Okay, I spent some time trying to do a miniature version of your set up and worked with another engineer on how to troubleshoot something like this. I had no issues scaling up and down services to newly added hosts and the load balancer eventually hit the new containers. 

1. How long did you wait before you started testing the global load balancer? Since you have so many internal ones pointing to the global one and the general scale of your application, it might take awhile for all the changes to get propogated.

2. If you see this next time, you can check the haproxy.cfg files on the global load balancer (to make sure it's still pointing to all your internal load balancers) and on each individual internal load balancer (to make it's pointing to all your service apps). By doing a `cat /etc/haproxy/haproxy.cfg`, you would be able to see all the container IPs programmed in the backend which should correspond to all the IPs in your application.

3. Just to clarify on the round robin algorithm, when you scale back up to 15 containers on each service, round robin will go through each service in mot likely the order of when the container was created before moving on to the next service. So, there is a chance that due to your scale, you might not have seen the containers on the newly added host?

And finally, if all the backends are programmed correctly, then this would more likely be a haproxy issue.

**Comment by gordontyler** [comment]
1. I waited maybe 5-10 minutes after the services were brought up.
2. I'll check the config.
3. My JMeter instance is running at ~180 requests per second, so I'm assuming that it should reach every container in a few minutes.

Perhaps related to this, I've notice that some of the load balancer instances in my stack silently failed last night. Rancher and Docker show them as running, there's no errors in the Docker log, but the app service behind the load balancer is receiving no requests. This is only happening for some load balancers.

My environment:
- 20 Docker hosts
- 40 "appN" services (e.g. app1, app2, app3) -- 20 of which are scaled to 40 instances, and 20 of which are scaled to 35 instances
- 40 "dbN" services (e.g. db1, db2, db3), scale 1 -- each appN service is linked to the corresponding dbN service.
- 40 "appN-lb" load-balancer services (e.g. app1-lb, app2-lb), scale 1 -- each appN-lb service load-balances requests to the corresponding appN service.
- 1 public-app-lb load-balancer service, which load-balances requests to all of the appN-lb services.

I'm restarting the internal and public load balancers now to see if that fixes the problem. If not, I'll check the haproxy configs for the ones that are not working.

**Comment by gordontyler** [comment]
It's taking a very long time to restart the appN-lb load balancers -- 40 minutes later only 26 have been restarted. The `rancher-compose start appN-lb` command is not printing any output, which is odd.

I noticed the following in the Rancher server log:

```
2015-11-11 16:42:14,629 ERROR [:] [] [] [] [ecutorService-2] [.p.c.v.i.ConfigItemStatusManagerImpl] Failed null, exit code [134] output [bash: ../nptl/sysdeps/unix/sysv/linux/x86_64/../fork.c:141: __libc_fork: Assertion `({ __typeof (self->tid) __value; if (sizeof (__value) == 1) asm volatile ("movb %%fs:%P2,%b0" : "=q" (__value) : "0" (0), "i" (__builtin_offsetof (struct pthread, tid))); else if (sizeof (__value) == 4) asm volatile ("movl %%fs:%P1,%0" : "=r" (__value) : "i" (__builtin_offsetof (struct pthread, tid))); else { if (sizeof (__value) != 8) abort (); asm volatile ("movq %%fs:%P1,%q0" : "=r" (__value) : "i" (__builtin_offsetof (struct pthread, tid))); } __value; }) != ppid' failed.
/var/lib/cattle/common/scripts.sh: line 62: 16276 Aborted                 (core dumped) ${CATTLE_HOME}/config.sh "$@"
]
```

And also some of these:

```
2015-11-11 16:42:28,598 ERROR [:] [] [] [] [ecutorService-9] [o.a.c.m.context.NoExceptionRunnable ] Uncaught exception org.jooq.exception.DataChangedException: Database record has been changed
```

The Rancher UI is having difficulty displaying the stack. It takes a long time to load and the Chrome browser tab is consuming up to a GB of memory. It frequently crashes the tab if I leave it open while making service changes using the rancher-compose command.

Sorry if I'm posting this in the wrong issues, it's becoming hard to figure out which errors are related to specific events and behaviors.

**Comment by gordontyler** [comment]
I just caught this in the Rancher server logs:

```
time="2015-11-11T17:06:31Z" level=info msg="Exiting rancher-compose-executor" version=v0.5.0-rc2
time="2015-11-11T17:06:33Z" level=info msg="Starting rancher-compose-executor" version=v0.5.0-rc2
time="2015-11-11T17:06:33Z" level=info msg="Initializing event router" workerCount=10
time="2015-11-11T17:06:43Z" level=fatal msg="Unable to start event router" error="Timed out waiting for transtion."
time="2015-11-11T17:06:44Z" level=info msg="Starting rancher-compose-executor" version=v0.5.0-rc2
time="2015-11-11T17:06:44Z" level=info msg="Initializing event router" workerCount=10
time="2015-11-11T17:06:44Z" level=info msg="Connection established"
time="2015-11-11T17:07:35Z" level=info msg="Exiting go-machine-service..."
time="2015-11-11T17:07:40Z" level=info msg="Setting log level" logLevel=info
time="2015-11-11T17:07:40Z" level=info msg="Starting go-machine-service..." gitcommit=3c74782
time="2015-11-11T17:07:51Z" level=error msg="Unable to start EventRouter" Err="Timed out waiting for transtion."
time="2015-11-11T17:07:51Z" level=info msg="Exiting go-machine-service..."
time="2015-11-11T17:07:52Z" level=info msg="Setting log level" logLevel=info
time="2015-11-11T17:07:52Z" level=info msg="Starting go-machine-service..." gitcommit=3c74782
time="2015-11-11T17:07:52Z" level=info msg="Initializing event router" workerCount=10
time="2015-11-11T17:07:52Z" level=info msg="Connection established"
```

**Comment by ibuildthecloud** [comment]
@gordontyler On the Rancher server, have you adjusted the Java heap?  For your setup I'd recommend you run with `-e JAVA_OPT="-Xmx4096m"` on the docker run command.  We will probably change the rancher/server start script to adapt the heap size, but for larger/active setups the default 1gb is usually not enough.

Some time post DockerCon I'd love to see your setup first hand.  We will be focusing a lot of effort on scalability and reliable then.  Seem like you have a good real world setup.

How many hosts are you running?

**Comment by gordontyler** [comment]
I'm already running the Rancher server with a `-Xmx8g` option. I ran into OOME a few weeks ago with the default config and it totally busted the Rancher database. I had to recreate it from scratch, but it's been running without OOME since then.

I have one VM -- 16GB RAM, 2 CPUs -- running the Rancher server container and 20 VMs -- 32GB RAM. 2 CPUs -- running my application containers. The Docker Engine on the Rancher serer VM is not registered in the Rancher environment itself, but the rest are.

My environment is essentially a testbed for our own appserver monitoring software, so it's not really a "live" production system. I spoke to David Getzler on the phone two weeks ago about our setup, so you could get some details from him. But I would be glad to discuss it in detail if you'd like.

**Comment by ibuildthecloud** [comment]
Awesome, thanks!  Also, just a heads up Docker 1.9 + Rancher + a lot of containers does not work well because of https://github.com/docker/docker/issues/17720.  You will probably see hosts go into "reconnecting."  1.8 is better, but once 1.9.1 comes out with the fix, the performance of 1.9.1 will be the best.

**Comment by gordontyler** [comment]
I'm still using Docker 1.8.3 on all the VMs. I contemplated upgrading to 1.9 but I figured that it was too risky before I had completed this current round of load testing.

**Comment by deniseschannon** [comment]
@ibuildthecloud would say, do not upgrade to 1.9 as it doesn't scale. please wait until at least 1.9.1

**Comment by deniseschannon** [comment]
@gordontyler When you have a non working internal load balancer before restarting, you should check the haproxy.cfg of the load balancer to see if it's still programmed.

**Comment by gordontyler** [comment]
More load balancer weirdness...

I needed to increase the CPU allocation for my Docker hosts. So I iteratively shutdown each host, increased it's CPU allocation in vSphere and then started the host up again.

After all of this was completed, and allowing for about 30 minutes of "settle time", I have a number of load balancer containers on various hosts that are stuck in a "restarting" state according to the Rancher UI -- but in actual fact those containers are not even running on the host when I check with Docker directly. I tried manually starting one of the containers using `docker start` but it hasn't updated the UI.

**Comment by gordontyler** [comment]
So the load balancers *eventually* finished starting/restarting. They show as running with CPU/etc. stats in the UI. 

However, 11 out of the 40 app services have not received any traffic from the load balancers since I restarted the hosts. I've waited 30 minutes for config to propogate but still nothing. I did a spot check on one of the app load balancers' haproxy.cfg and it appears to have a valid list of server IP addresses. I also checked the public load balancer's haproxy.cfg and it appears to have a valid list of the internal app load balancer IP addresses.

I'm attempting to remove and recreate the broken app load balancers.
